### DISCUSSION TITLE
关于 UI 模块 之 WebView

### DISCUSSION THREAD

**Discussion by julywind168** [ejoy/ant#90] [Ideas]
## 嵌入一个 webkit 可能太重了，替代方案是 WebView
- [webview](https://github.com/webview/webview) 这个比较简单，但是不适合嵌入引擎
- 比较出名的tauri 也采用这种方案 [tauri-apps/wry](https://github.com/tauri-apps/wry) 功能更丰富，我用它和 bevy 集成成功。
- 之后我用 swift 为 Ant Engine 实现了 Macos 上的集成，仅用了150行代码
- 缺点可能是依赖玩家本机的 runtime, 根据 [微软的说法](https://blogs.windows.com/msedgedev/2022/12/14/delivering-microsoft-edge-webview2-runtime-to-managed-windows-10-devices/)，大部分 win10 已经有了webview2, win11 则内置。
#### 是否会考虑集成 WebView 作为一个 UI 模块的备选项呢 （by the way 这个也可以替代 imgui ）

**Comment by cloudwu** [comment]
这些方案对于游戏来说性能堪忧，且没有优化余地。游戏里 ui 应该占非常小的比例，而不能反客为主，把 cpu 和内存都让给 ui 。

游戏也不需要把 ui 和游戏逻辑完全隔离开。这些方案都会独享 字体/贴图/着色器 ，等于本应该只有一份的东西在游戏内变成两份。

难以将 3d 渲染融入 ui 。

js vm 有巨大的内存及 cpu 开销。且和 lua vm 有更多的的数据交换代价。

即使不跑游戏，只跑 ui ，就像现在很多 web app 那样，纯 js 框架性能都是问题。

**Reply by julywind168** [reply]
ui 跑独立线程的情况下，cpu 的占用和游戏就没太大关系了

以 https://tauri.app/ 为例，这个网页 cpu 大概是 5%, 内存 200mb 左右波动

内存吃的确实有点多，但只考虑 pc 平台，这些都还可以接受吧。（开发  mmo 端游中）

3d 模型渲染的话，ui 上开个口子(div 背景透明) 让游戏把模型渲染到该 div 位置就可以了吧

主要是 ant 现在开发 ui 太难了，具体来说缺乏一套 ui 组件库。一些 web 中非常基础的 进度条 下拉框 等等 都得手撸，更无法和一些 js ui 组件库比了。

建议引擎为开发自定义控件提供一些方便，并建一个 ui components 仓库，通过社区力量慢慢完善这方面的体验。

可以预见是，这需要非常长的时间。所以我还是考虑自己集成 webview 来开发 ui (only macos & windows)

**Comment by julywind168** [comment]
```
  /// Sets whether the WebView should be transparent.
  ///
  /// ## Platform-specific:
  ///
  /// **Windows 7**: Not supported.
  pub fn with_transparent(mut self, transparent: bool) -> Self {
    self.attrs.transparent = transparent;
    self
  }
```
算了，不折腾了。看 wry 的接口 win7 好像弄不了透明背景 😓

**Comment by actboy168** [comment]
react-native存在的原因不就是webview性能不行么，哪怕只是非游戏的应用。

**Comment by vkensou** [comment]
只是为了UI的话没必要吧。。。哪怕集成下fairygui呢？甚至上古的mygui、cegui也行啊。。

**Comment by constly** [comment]
我在fairygui社区发了一个[帖子](https://ask.fairygui.com/?/question/27050)，希望他们能考虑对接下

**Reply by danggui1995** [reply]
顶fairygui一票，unity 6年使用者表示真的很好用
